sql/mdl.h

```diff
@@ -934,18 +887,13 @@ template <typename T, typename K> class hash_trait
   {
     return lhs == rhs;
   }
-  static bool is_equal(const elem_type &lhs, const K &rhs)
-  {
-    return lhs->get_key()->is_equal(rhs.mdl_key) &&
-           lhs->has_stronger_or_equal_type(rhs.type);
-  }
 
   static bool is_empty(const elem_type &el) { return el == nullptr; }
   static void set_null(elem_type &el) { el= nullptr; }
 
   static my_hash_value_type get_hash_value(T *t)
   {
-    return t->get_key()->tc_hash_value();
+    return MDL_key_trait<T>::get_hash_value_from_key(t->get_key());
```

Comment:
This way you're making a class named `hash_trait` bound to MDL_key_trait

Reply:
It really is. Are you not happy with this way, or are you just suggesting renaming the class?

Reply:
Suppose we'll have another class, that would be stored in our hash, it can be a table of plugins for example. 
The developer would study the usage of our hash, make their own key trait (since plugins don't use MDL keys), and try to reuse hash_trait, but they would fail because of this dependency in `get_hash_value`.

Reply:
my suggestion is to leave `get_key()` here and call `get_hash_value_from_key` (also please rename it to just `get_hash_value`) from the hash:
```cpp
template <class key_trait> // since we don't know the exact return type and we don't have `auto` here
static const key_trait::key_type& get_key(const T *value)
{
  return value->get_key();
}
```

Now in the hash code:
```cpp
  my_hash_value_type get_hash_value(const T *value)
  { return key_trait::get_hash_value(value_trait::get_key(value)); }
```

Note that `key_trait::get_hash_value` should accept const key_type& in the way i wrote it here. 
You can also rephrase it in pointers, if you like. But `const` is mandatory.